### ISSUE TITLE
Planning for unchangeable topics

### ISSUE THREAD

**Issue by robdejonge** [homieiot/convention#107]
All new into the world of home automation, MQTT especially is appealing to me and I've spent a few days trying to come up with a solid architecture. Until I read about Homie, and I was sold. The only problem for which I've not seen cover, is the transient nature of sensors and devices in a home automation setup. 

Today, I use sensor A to monitor the temperature in my living room. Tomorrow, I might see a fancier version B and want to replace it. Perhaps sensor A used a specific topic to publish on, defined by the vendor, and version B now uses an entirely different topic. If that were the case, I think adhering to Homie would mean I have to update the topic name in maybe a dozen different places to make sure the automation continues. 

What I was planning to do is to map one topic to another. For example, ShenzenElectronicsSensor55B/temperature gets mapped to home/livingroom/temperature. The automation subscribes to the latter. When I replace that sensor for HW/3723/temperature broadcasts from another hardware company, everything else will continue to work. 

Excuse my ignorance if this is covered somehow, but I thought I'd raise it nonetheless.

**Comment by lorenwest** [comment]
Hi Rob - welcome to the world of IoT and automation! 

From a Homie perspective, a company either presents their product using the Homie convention, or they don't. A good product will let you tell it the name of the topic. Not sure if that's a part of the convention, but the ESP8266 implementation has this feature. Products of this caliber will let you configure their topic, so you won't need a mapping.

Then there's products that don't conform to the Homie convention. If they happen to let you set the topic they publish on, that'll save some translation, but it still won't make them Homie compatible.

Homie is much more than describing topic names. It's a discovery mechanism, and a way to differentiate a request for changing state from the expression of state provided by the device. No changes to conventions will make a non-Homie product into a Homie product.

Your examples are of products that aren't Homie compatible. In this case, to make them Homie compatible, you'll need to write some software running on the MQTT bus that reads state and expresses it on a different topic - one that's Homie compatible. That requires the software to run somewhere - usually on an automation engine like [Node-Red](https://nodered.org/).

In short, changing Homie convention isn't enough to translate a non-Homie device into a Homie-compatible deice. It takes a software translation layer that understands both the device specifics and the Homie convention.

Hope that helps.

**Comment by robdejonge** [comment]
Thanks, that clarifies the intent of Homie quite a bit.
